File: kotlin/advent/src/main/kotlin/org/gristle/pdxpuzzles/advent/y2015/Y15D08.kt
package org.gristle.pdxpuzzles.advent.y2015

import org.gristle.pdxpuzzles.advent.utilities.Day

class Y15D08(input: String) : Day {
    private val lines = input.lines()

    private val replaceRx = """\\\\|\\"|\\x[\da-f]{2}""".toRegex()

    private fun String.charsInMemory(): Int {
        return drop(1).dropLast(1).replace(replaceRx, "X").length
    }

    private fun String.encodedLength() = length + 2 + count { it in """\"""" }

    private val totalLength = lines.sumOf(String::length)

    override fun part1() = totalLength - lines.sumOf { it.charsInMemory() }

    override fun part2() = lines.sumOf { it.encodedLength() } - totalLength
}

fun main() = Day.runDay(Y15D08::class)

//    Class creation: 23ms
//    Part 1: 1333 (10ms)
//    Part 2: 2046 (1ms)
//    Total time: 36ms

File: rust/advent/src/bin/2015/y15d08.rs
use advent::utilities::get_input::get_input;
use lazy_regex::regex;
use utilities::structs::stopwatch::{ReportDuration, Stopwatch};

type Input<'a> = &'a str;
type Output = usize;

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let input = get_input(15, 8).unwrap();
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(&input), stopwatch.lap().report());
    println!("2. {} ({})", part2(&input), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

fn chars_in_memory(s: &str) -> usize {
    let rx = regex!(r#"\\\\|\\"|\\x[\da-f]{2}"#);
    rx.replace_all(&s[1..s.len() - 1], "X").len()
}

fn encoded_length(s: &str) -> usize {
    s.len() + 2 + s.as_bytes().iter()
        .filter(|&&b| b == b'"' || b == b'\\')
        .count()
}

fn part1(input: Input) -> Output {
    let total_length = input.len() - input.as_bytes().iter()
        .filter(|&&b| b == b'\n')
        .count();
    total_length - input.lines()
        .map(|line| chars_in_memory(line))
        .sum::<usize>()
}

fn part2(input: Input) -> Output {
    let total_length = input.len() - input.as_bytes().iter()
        .filter(|&&b| b == b'\n')
        .count();
    input.lines()
        .map(|line| encoded_length(line))
        .sum::<usize>() - total_length
}

#[test]
fn default() {
    let input = get_input(15, 8).unwrap();
    assert_eq!(1333, part1(&input));
    assert_eq!(2046, part2(&input));
}

// Input parsed (42μs)
// 1. 1333 (385μs)
// 2. 2046 (16μs)
// Total: 445μs
